src/components/RowItem.tsx

```diff
@@ -34,14 +35,16 @@ function RowItem<T>(props: Props<T>) {
     drag(itemKey);
   }, []);
 
-  const { renderItem, item, itemKey, extraData } = props;
+  const { renderItem, item, itemKey, extraData, passIndexToRenderItem } = props;
   return (
     <MemoizedInner
       isActive={activeKey === itemKey}
       drag={drag}
       renderItem={renderItem}
       item={item}
-      index={keyToIndexRef.current.get(itemKey)}
+      index={
```

Comment:
Thanks for the PR! I'll have to think about whether this optimization belongs in the package or in the client code.

Is this any more performant than omitting `index` by memoizing within  your own `renderItem`?

something like:

```tsx
const renderItem = useCallback(({ item, drag }) => {
  return <MyMemoizedItem item={item} drag={drag} />
}, [])

...

const MyMemoizedItem = React.memo(({ item, drag }) => { ... })
```

Reply:
We tried that, and it didn't work. This is because when I hovered over why the child components rerendered, it said that not only the props changed, but that the whole item was new, starting from the child of `Inner` in `RowItem.tsx`

The render tree looked a bit like this for each list item:

- RowItem (no changes)
- MemoizedInner (no changes)
- anonymous render function (newly created)
- MyChildComponent (newly created)

I suspect this is because `renderItem` isn't a typical react component and rather is just a render function.

This was the best fix we tried that worked at the time. One alternative fix that may work may be to get rid of `MemoizedInner` and `Inner` and just do the child rendering inside of `RowItem` (which would be equivalent to the `CellRenderer` in VirtualizedLists), but I'm not sure if that'd work.

Reply:
I see, that makes sense. I have a PR that is a massive upgrade to this package (uses reanimated v2 api, gesture handler v2), I've updated the api so that `renderItem` gets a `getIndex()` function that never changes identity instead of an `index` prop: https://github.com/computerjazz/react-native-draggable-flatlist/pull/335/files#diff-4e6ce7938fbbf1a379a2275894c898fbaf661b4699089774369e712e32520057R40-R51

I think that should do the same thing? Feel free to try it out by installing the beta version: https://www.npmjs.com/package/react-native-draggable-flatlist?activeTab=versions

Reply:
@computerjazz that is fantastic, super cool as I just ran into exactly this issue trying to use layout animations with a draggable flatlist. Any ETA on when the version 4 will exit beta? Are you close or should we think months away?

Reply:
If you could let me know why it crashed that would be useful. index -> getIndex is the main API change

Reply:
@computerjazz of course, mind you I just upgraded and didnt make any changes. I don't see where I should need to use getIndex instead of index, as I am using keyExtractor in v3. Here was the error, seems like some zIndex that is NaN:

```
Invariant Violation: [2297,"RCTView",391,{"flexDirection":"row","backgroundColor":4294967295,"width":390,"paddingVertical":3.361764705882353,"paddingLeft":19.5,"paddingBottom":5.602941176470589,"zIndex":"<<NaN>>"}] is not usable as a native method argument

This error is located at:
    in RCTView (created by View)
    in View
    in Unknown (created by Inner)
    in RCTView (created by View)
    in View (created by AnimatedComponent)
    in AnimatedComponent
    in AnimatedComponentWrapper (created by TouchableOpacity)
    in TouchableOpacity (created by TouchableOpacity)
    in TouchableOpacity (created by Inner)
    in Inner (created by RowItem)
    in RowItem (created by CellRenderer)
    in CellProvider (created by CellRendererComponent)
    in RCTView (created by View)
    in View (created by AnimatedComponent(View))
    in AnimatedComponent(View)
    in Unknown (created by CellRendererComponent)
    in CellRendererComponent (created by CellRenderer)
    in VirtualizedListCellContextProvider (created by CellRenderer)
    in CellRenderer (created by VirtualizedList)
    in RCTScrollContentView (created by ScrollView)
    in RCTScrollView (created by ScrollView)
    in ScrollView (created by ScrollView)
    in ScrollView (created by NativeViewGestureHandler)
    in NativeViewGestureHandler (created by ScrollView)
    in ScrollView (created by VirtualizedList)
    in VirtualizedListContextProvider (created by VirtualizedList)
    in VirtualizedList (created by FlatList)
    in FlatList
    in Unknown (created by AnimatedComponent(Component))
    in AnimatedComponent(Component)
    in Unknown (created by DraggableFlatListInner)
    in RCTView (created by View)
    in View (created by AnimatedComponent(View))
    in AnimatedComponent(View)
    in Unknown (created by DraggableFlatListInner)
    in Wrap (created by AnimatedComponent(Wrap))
    in AnimatedComponent(Wrap)
    in Unknown (created by GestureDetector)
    in GestureDetector (created by DraggableFlatListInner)
    in DraggableFlatListProvider (created by DraggableFlatListInner)
    in DraggableFlatListInner
    in RefProvider
    in AnimatedValueProvider
    in PropsProvider
    in DraggableFlatList (created by IngredientInput)
    in RCTView (created by View)
    in View (created by IngredientInput)
    in IngredientInput (created by IngredientsInputScreen)
    in RCTView (created by View)
    in View (created by InputRow)
    in InputRow (created by IngredientsInputScreen)
    in RCTView (created by View)
    in View (created by IngredientsInputScreen)
    in RCTScrollContentView (created by ScrollView)
    in RCTScrollView (created by ScrollView)
    in ScrollView (created by ScrollView)
    in ScrollView (created by KeyboardAwareScrollView)
    in KeyboardAwareScrollView (created by IngredientsInputScreen)
    in Suspense (created by IngredientsInputScreen)
    in RNCSafeAreaView
    in Unknown (created by IngredientsInputScreen)
    in IngredientsInputScreen (created by SceneView)
    in StaticContainer
    in EnsureSingleNavigator (created by SceneView)
    in SceneView (created by CardContainer)
    in RCTView (created by View)
    in View (created by CardContainer)
    in RCTView (created by View)
    in View (created by CardContainer)
    in RCTView (created by View)
    in View
    in CardSheet (created by Card)
    in RCTView (created by View)
    in View (created by AnimatedComponent)
    in AnimatedComponent
    in AnimatedComponentWrapper (created by PanGestureHandler)
    in PanGestureHandler (created by PanGestureHandler)
    in PanGestureHandler (created by Card)
    in RCTView (created by View)
    in View (created by AnimatedComponent)
    in AnimatedComponent
    in AnimatedComponentWrapper (created by Card)
    in RCTView (created by View)
    in View (created by Card)
    in Card (created by CardContainer)
    in CardContainer (created by CardStack)
    in RNSScreen (created by AnimatedComponent)
    in AnimatedComponent
    in AnimatedComponentWrapper (created by Screen)
    in MaybeFreeze (created by Screen)
    in Screen (created by MaybeScreen)
    in MaybeScreen (created by CardStack)
    in RNSScreenContainer (created by ScreenContainer)
    in ScreenContainer (created by MaybeScreenContainer)
    in MaybeScreenContainer (created by CardStack)
    in RCTView (created by View)
    in View (created by Background)
    in Background (created by CardStack)
    in CardStack (created by HeaderShownContext)
    in RCTView (created by View)
    in View (created by SafeAreaInsetsContext)
    in SafeAreaProviderCompat (created by StackView)
    in RCTView (created by View)
    in View (created by GestureHandlerRootView)
    in GestureHandlerRootView (created by StackView)
    in StackView (created by StackNavigator)
    in Unknown (created by StackNavigator)
    in StackNavigator (created by RecipeWizardStack)
    in RecipeWizardStack (created by SceneView)
    in Suspense (created by SceneView)
    in StaticContainer
    in EnsureSingleNavigator (created by SceneView)
    in SceneView (created by CardContainer)
    in RCTView (created by View)
    in View (created by CardContainer)
    in RCTView (created by View)
    in View (created by CardContainer)
    in RCTView (created by View)
    in View
    in CardSheet (created by Card)
    in RCTView (created by View)
    in View (created by AnimatedComponent)
    in AnimatedComponent
    in AnimatedComponentWrapper (created by PanGestureHandler)
    in PanGestureHandler (created by PanGestureHandler)
    in PanGestureHandler (created by Card)
    in RCTView (created by View)
    in View (created by AnimatedComponent)
    in AnimatedComponent
    in AnimatedComponentWrapper (created by Card)
    in RCTView (created by View)
    in View (created by Card)
    in Card (created by CardContainer)
    in CardContainer (created by CardStack)
    in RNSScreen (created by AnimatedComponent)
    in AnimatedComponent
    in AnimatedComponentWrapper (created by Screen)
    in MaybeFreeze (created by Screen)
    in Screen (created by MaybeScreen)
    in MaybeScreen (created by CardStack)
    in RNSScreenContainer (created by ScreenContainer)
    in ScreenContainer (created by MaybeScreenContainer)
    in MaybeScreenContainer (created by CardStack)
    in RCTView (created by View)
    in View (created by Background)
    in Background (created by CardStack)
    in CardStack (created by HeaderShownContext)
    in RCTView (created by View)
    in View (created by SafeAreaInsetsContext)
    in SafeAreaProviderCompat (created by StackView)
    in RCTView (created by View)
    in View (created by GestureHandlerRootView)
    in GestureHandlerRootView (created by StackView)
    in StackView (created by StackNavigator)
    in Unknown (created by StackNavigator)
    in StackNavigator (created by HomeStack)
    in HomeStack (created by SceneView)
    in StaticContainer
    in EnsureSingleNavigator (created by SceneView)
    in SceneView (created by Drawer)
    in RCTView (created by View)
    in View (created by Screen)
    in RCTView (created by View)
    in View (created by Background)
    in Background (created by Screen)
    in Screen (created by Drawer)
    in RNSScreen (created by AnimatedComponent)
    in AnimatedComponent
    in AnimatedComponentWrapper (created by Screen)
    in MaybeFreeze (created by Screen)
    in Screen (created by MaybeScreen)
    in MaybeScreen (created by Drawer)
    in RNSScreenNavigationContainer (created by ScreenContainer)
    in ScreenContainer (created by MaybeScreenContainer)
    in MaybeScreenContainer (created by Drawer)
    in RCTView (created by View)
    in View (created by Drawer)
    in RCTView (created by View)
    in View (created by AnimatedComponent(View))
    in AnimatedComponent(View)
    in Unknown (created by Drawer)
    in RCTView (created by View)
    in View (created by AnimatedComponent(View))
    in AnimatedComponent(View)
    in Unknown (created by PanGestureHandler)
    in PanGestureHandler (created by Drawer)
    in Drawer (created by DrawerViewBase)
    in DrawerViewBase (created by DrawerView)
    in RCTView (created by View)
    in View (created by GestureHandlerRootView)
    in GestureHandlerRootView (created by DrawerView)
    in RCTView (created by View)
    in View (created by SafeAreaInsetsContext)
    in SafeAreaProviderCompat (created by DrawerView)
    in DrawerView (created by DrawerNavigator)
    in Unknown (created by DrawerNavigator)
    in DrawerNavigator (created by Main)
    in Main (created by SceneView)
    in StaticContainer
    in EnsureSingleNavigator (created by SceneView)
    in SceneView (created by CardContainer)
    in RCTView (created by View)
    in View (created by CardContainer)
    in RCTView (created by View)
    in View (created by CardContainer)
    in RCTView (created by View)
    in View
    in CardSheet (created by Card)
    in RCTView (created by View)
    in View (created by AnimatedComponent)
    in AnimatedComponent
    in AnimatedComponentWrapper (created by PanGestureHandler)
    in PanGestureHandler (created by PanGestureHandler)
    in PanGestureHandler (created by Card)
    in RCTView (created by View)
    in View (created by AnimatedComponent)
    in AnimatedComponent
    in AnimatedComponentWrapper (created by Card)
    in RCTView (created by View)
    in View (created by Card)
    in Card (created by CardContainer)
    in CardContainer (created by CardStack)
    in RNSScreen (created by AnimatedComponent)
    in AnimatedComponent
    in AnimatedComponentWrapper (created by Screen)
    in MaybeFreeze (created by Screen)
    in Screen (created by MaybeScreen)
    in MaybeScreen (created by CardStack)
    in RNSScreenContainer (created by ScreenContainer)
    in ScreenContainer (created by MaybeScreenContainer)
    in MaybeScreenContainer (created by CardStack)
    in RCTView (created by View)
    in View (created by Background)
    in Background (created by CardStack)
    in CardStack (created by Head...(truncated to the first 10000 characters)
at node_modules/react-native/Libraries/Core/ExceptionsManager.js:95:4 in reportException
- ... 25 more stack frames from framework internals
```

Reply:
Thanks! Hm that style doesn't look like it's coming from within the draggable flatlist component -- are you driving any `zIndex` animations using values from within the package? (i.e. `useOnCellActiveAnimation()` or other values received as props )?

Reply:
@computerjazz I see, nice one. I had used "index" in a zIndex, but after moving to using the function `getIndex` it works fine now. The only thing I noticed was that, when dragging an item, the "isActive" seem to light up the previous item (index wise) before it moves to the ending item when you drop the item after drag. See gif below for better explanation:

![Simulator Screen Recording - iPhone 13 Pro - 2022-06-20 at 19 57 12](https://user-images.githubusercontent.com/24279217/174656781-dafca94f-fb2b-49a6-b912-0c097edaa738.gif)

Reply:
@computerjazz initial load of draggable flatlist with more items (e.g. 8 or so), feels a little bit slower. Can that be so?

Reply:
@computerjazz same speed as old one, I checked. Another thing I noticed, the drag gesture from parent scrollview are ignored. So if you have a long draggable flatlist nested inside a scrollview - you cant scroll any longer with the parent scrollview. I assume there is some "simultaneous handler" setting that needs setting up, just thought I would mention as it is a difference vs. v3. 

Really nice work otherwise!!

Reply:
doh, that end-of-drag flicker is the most difficult thing to solve about this whole package (draggable flatlist works by swapping out the modified list for the next "actual" list in a single frame -- if there are any timing issues a flicker happens). I thought I solved it, but if you could provide a reproduceable snack or repo i'd really appreciate it. Different versions of reanimated also may have different behavior -- i've been testing with 2.8.0.

As far as nested scrollviews go -- are you using `NestableDraggableFlatlist` and `NestableScrollContainer`?

Reply:
@computerjazz hmm, I can't reproduce it in a separate target repo or snack. It must be something in my main app, that creates the condition for it to happen in v4 and not v3. I can invite you to it so you can spin it up yourself, but will be very heavy compared to a targeted repo. :/

I am also on reanimated 2.8.0 and react-native 0.68.2

Reply:
sure, if it doesn't require any local backend/api keys/etc i'd be happy to take a look.

Reply:
I was able to reproduce the flickering _(not with this particular PR, but with an "official build")_ using Realm DB data to render my items. In this issue I've posted some examples:

https://github.com/computerjazz/react-native-draggable-flatlist/issues/434#issuecomment-1464981428